tests/test_2125_type_of_scalar.py

```diff
@@ -0,0 +1,30 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import numpy as np
+import pytest  # noqa: F401
+
+import awkward as ak
+
+
+def test():
+    assert ak.type(np.int16()) == ak.types.ScalarType(ak.types.NumpyType("int16"))
+    assert ak.type(np.uint32) == ak.types.ScalarType(ak.types.NumpyType("uint32"))
+    assert ak.type(np.dtype("complex128")) == ak.types.ScalarType(
+        ak.types.NumpyType("complex128")
+    )
```

Comment:
Hmmmm. The same argument I made for `ak.type(np.uint32)` could be made for `ak.type(np.dtype(...))`, although the latter is necessary to express units on datetimedeltas.

~~Actually, I think the same argument holds. Even though `np.dtype(...)` is a Python instance, it represents a NumPy type, just as `ak.types.Type` instances represent Awkward types. I don't see how it benefits anyone for the `ak.type` function to accept NumPy or Awkward _types_.~~

Actually actually, I changed my mind. Even though it is a category error (like `raise NotImplementedError` without the `()`), it's a useful one (again, like `raise NotImplementedError`). This usage of the `ak.type` function is to find out how we would name a type in the Awkward system: "`np.dtype("complex128")`" is NumPy's name for it, and "`ak.types.NumpyType("complex128")`" is ours. Similarly, one can use `ak.type` to find out how Python `int` and `str` are mapped to Awkward types through `ak.from_iter`, right?

```python
>>> ak.type(int)
...
TypeError: while calling

    ak.type(
        array = type-instance
        behavior = None
    )

Error details: unrecognized array type: <class 'int'>
>>> ak.type(str)
...
TypeError: while calling

    ak.type(
        array = type-instance
        behavior = None
    )

Error details: unrecognized array type: <class 'str'>
```

Well, okay, it can't. But also in trying it out, I found

```python
>>> ak.type("int32")
ArrayType(NumpyType('uint8', parameters={'__array__': 'char'}, typestr='char'), 5)
```

If users are expecting to be able to pass any dtype-like as an argument to `ak.type`, they'll be surprised by this one.

So, actually actually actually, I think `ak.type` should not accept a type as its argument, even if that's useful as a way of regularizing types "in the wild" into Awkward-style types. It's conflating too many different things, and we get into territory where two mutually incompatible expectations are relevant: arguments that accept a dtype often accept a dtype-like, which includes strings, and a string is an array-like with the above type. We wouldn't be able to make this work as expected for two users, each of whom has one or the other of these expectations.

So I think `ak.type` should only accept array-like arguments, not type-like arguments. That's where I've finally landed after flopping back and forth.